File: Machine_Learning/Regression/Support_Vector_Machines/SupportVectorMachines.py
#!/usr/bin/env python
# coding: utf-8

# # SVM in Python for Classification Problem using Sklearn - linear

# ## Use the sklearn module to create Linear model.

# In[1]:


import sklearn
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import mlxtend


# ## Load the data into Python enviorment (Iris)

# In[2]:


from sklearn.datasets import load_iris
iris = load_iris()
X = pd.DataFrame(iris.data, columns=iris.feature_names) #Input
y = iris.target #Output


# ## Split the Data

# In[3]:


from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=1)


# ## Train the Model

# In[4]:


from sklearn import svm
# Design the model
model = svm.SVC(kernel='linear') #SVM Linear 
model.fit(X_train,y_train)


# ## Test the Model

# In[5]:


from sklearn.metrics import confusion_matrix
# Test result
y_predict = model.predict(X_test)
confusion_matrix(y_test, y_predict) #No mistake


# ## Capture the Accuracy of the Model

# In[6]:


from sklearn.metrics import accuracy_score
accuracy_score(y_test, y_predict)*100


# ## Get the Number of Support vectors for Each Class

# In[7]:


model.n_support_


# # SVM in Python for Classification using Sklearn - ploy

# ## Fit the model

# In[8]:


clf = svm.SVC(kernel='poly', degree=3) #degree = 3 for the number of factors
clf.fit(X_train, y_train)


# ## Test the model.

# In[9]:


y_predict = clf.predict(X_test)
confusion_matrix(y_test, y_predict)


# In[10]:


accuracy_score(y_test, y_predict)*100


# # SVM in Python for Classification using sklearn - rbf

# ## Train the model.

# In[11]:


clf = svm.SVC(kernel='rbf', gamma=.3) #gamma = .01* number of factor
clf.fit(X_train, y_train)


# ## Test the model.

# In[12]:


y_predict = clf.predict(X_test)
confusion_matrix(y_test, y_predict)


# In[13]:


accuracy_score(y_test, y_predict)*100


# # Grid Search SVM Methods using sklearn and GridSearchCV

# In[18]:


from sklearn.model_selection import GridSearchCV
parameters = {'kernel':('linear', 'rbf', 'poly'), 'C':[0.1, 0.5, 1, 10], 'degree':[2,3,4]}
svc = svm.SVC()
clf = GridSearchCV(svc, parameters, cv=5)
clf.fit(X_train, y_train)


# # Select the Best Method and Parameters

# In[19]:


clf.best_params_ #Best models. Parameters should be used to retrain "poly" model (the best model)

File: Machine_Learning/Regression/Support_Vector_Machines/SupportVectorMachines.R
#SVM Linear in R using caret
#Use caret to implement SVM using a different kernels in R to fit regression type problems (Google close price)
library(caret)

#Load data into R enviorment
library(quantmod)
start <- as.Date(Sys.Date()-(365*5))
end <- as.Date(Sys.Date())
getSymbols("GOOG", src = "yahoo", from = start, to = end)

data = GOOG
colnames(data) = c("Open", "High", "Low", "Close", "Volume", "Adjusted")

#Visualize the Data 
plot(data[, "Close"], main = "Close Price") #Close Price is linear

#Create Test Harnesses
control <- trainControl(method="cv", number=10) #10 folds and R2 since regression problem
metric <- "Rsquared"

#Split the Data
split<-createDataPartition(y = data$Close, p = 0.7, list = FALSE)
train<-data[split,]
valid<-data[-split,]

#Build a SVM Model for Regression problem using svmLinear
set.seed(7)
fit.LM <- train(Close~., data = train, method = "svmLinear", trControl=control, metric=metric)

#Summarize the Results Briefly
fit.LM
#R2 is good. However RMSE is not ($39.89 off of the real close price which is a big error rate). Linear regression is better than SVM for regression problems

#Summarize the Entire Final Model
summary(fit.LM$finalModel)

#Create Prediction using Trained SVM #Post resample technique for regression problems
predictedValues<-predict(fit.LM, valid)
modelvalues<-data.frame(obs = valid$Close, pred=predictedValues)
postResample(pred = predictedValues, obs = valid$Close) #We got a good R2 (0.99959). RMSE and MAE are very off. It is not a good model to predict actuals.



##SVM Non-Linear in R using caret

#Use caret to implement SVM using different kernels in R to fit classification problems.
library(caret)

#Load data into R enviroment
data("iris")
iris = iris
str(iris)

#Visualization (EDA)

#Split the Data (70/30 Split)
index = createDataPartition(iris[,1], p =0.70, list = FALSE)
dim(index)

#Training Data
training = iris[index,]
dim(training)

#Validation Data
valid = iris[-index,]
dim(valid)

#Create Test Harnesses
control <- trainControl(method="cv", number=10)
metric <- "Accuracy"

#Build a SVM Model for Classification using svmLinear
set.seed(7)
#fit.LR <- train(Species~., data = training, method = "svmLinear", family=binomial(), trControl=control, metric=metric)
fit.LR <- train(Species~., data = training, method = "svmLinear", trControl=control, metric=metric) #family=binomial() used for binomial problems

#Summarize the Results Briefly
fit.LR #Good accuracy 0.9809091

#Summarize the Entire Final Model
summary(fit.LR$finalModel)

#Create Prediciton using Trained SVM
data.pred = predict(fit.LR, newdata = valid)
table(data.pred, valid$Species)

#Check Error
error.rate = round(mean(data.pred != valid$Species,2))
error.rate

#Confusion Matrix
cm = confusionMatrix(as.factor(data.pred), reference = as.factor(valid$Species), mode = "prec_recall")
print(cm) #good 0.9302 with CI (0.8094, 0.9854)). Setosa is handled easily. Few errors done on versicolor and virginica


##Build a SVM Model for Classification using svmRadial
set.seed(7)
fit.svmRadial <- train(Species~., data = training, method="svmRadial", metric=metric, trControl=control)

#Summarize the Results Briefly
fit.svmRadial #Doing penalization matrix. 0.5 is the best with accuracy of 0.9727273

#Summarize the Entire Final Model
summary(fit.svmRadial$finalModel)

#Create Prediciton using Trained Regression
data.pred = predict(fit.svmRadial, newdata = valid)
table(data.pred, valid$Species)

#Check Error
error.rate = round(mean(data.pred != valid$Species,2))
error.rate

#Confusion Matrix
cm = confusionMatrix(as.factor(data.pred), reference = as.factor(valid$Species), mode = "prec_recall")
print(cm) # Accuracy : 0.907  &  95% CI : (0.7786, 0.9741)

##Build a SVM Model for Classification using svmPoly
set.seed(7)
fit.svmPoly <- train(Species~., data = training, method="svmPoly", metric=metric, trControl=control)

#Summarize the Results Briefly
fit.svmPoly

#Summarize the Entire Final Model
summary(fit.svmPoly$finalModel)

#Create Prediciton using Trained Regression
data.pred = predict(fit.svmPoly, newdata = valid)
table(data.pred, valid$Species)

#Check Error
error.rate = round(mean(data.pred != valid$Species,2))
error.rate

#Confusion Matrix
cm = confusionMatrix(as.factor(data.pred), reference = as.factor(valid$Species), mode = "prec_recall")
print(cm) #Accuracy : 0.9302  & 95% CI : (0.8094, 0.9854)

#Compare SVMs Models
results = resamples(list(svmLinear = fit.LR, svmRadial=fit.svmRadial, svmPoly = fit.svmPoly))
summary(results)


#Visualize Comparison
dotplot(results) #Poly trained the best. And handled versicolor and virginica classification the best.



